scan-command/src/main/java/io/ballerina/scan/internal/StaticCodeAnalyzer.java

```diff
@@ -100,6 +116,77 @@ public void visit(ImplicitAnonymousFunctionExpressionNode implicitAnonymousFunct
         this.visitSyntaxNode(implicitAnonymousFunctionExpressionNode.expression());
     }
 
+    @Override
+    public void visit(ClassDefinitionNode classDefinitionNode) {
+        checkNonIsolatedPublicClassDefinitions(classDefinitionNode);
+        checkNonIsolatedPublicClassMembers(classDefinitionNode);
+    }
+
+    private void checkNonIsolatedPublicClassDefinitions(ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof ClassSymbol classSymbol) {
+                List<Qualifier> qualifiers = classSymbol.qualifiers();
+                if (getQualifier(qualifiers, PUBLIC_KEYWORD.stringValue()) &&
+                        !getQualifier(qualifiers, ISOLATED_KEYWORD.stringValue())) {
+                    reportIssue(classDefinitionNode, CoreRule.PUBLIC_NON_ISOLATED_CONSTRUCT);
+                }
+            }
+        });
+    }
+
+    private void checkNonIsolatedPublicClassMembers(ClassDefinitionNode classDefinitionNode) {
+        classDefinitionNode.members().forEach(member -> {
+            semanticModel.symbol(member).ifPresent(symbol -> {
+                if (symbol.kind() == SymbolKind.METHOD) {
+                    checkNonIsolatedPublicMethods((FunctionDefinitionNode) member, classDefinitionNode);
+                }
+                member.accept(this);
+            });
+        });
+    }
+
+    private void checkNonIsolatedPublicMethods(FunctionDefinitionNode member,
+                                               ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof ClassSymbol classSymbol) {
```

Comment:
Aren't you repeating this bit unnecessarily foe each member?

Reply:
Do you mean to directly cast the object without checking?